### ISSUE TITLE
pygtk2 app fails to load GIF images from pyinstaller exe

### ISSUE THREAD

**Issue by yellowtailfan** [Alexpux/MINGW-packages#3374]
I think your recent changes in #3339 broke image loading in pygtk2 when run from an exe built using pyinstaller under MinGW-w64 on Windows 10.

An app was working fine under pyinstaller using mingw-w64-x86_64-gdk-pixbuf2 2.36.11-1. Then yesterday I ran pacman -Syuu, which included mingw-w64-x86_64-gdk-pixbuf2 2.36.11-2 and the app broke with this error:
```
glib.GError: Unrecognized image file format
```

The exception was raised when the script creates a:
```
gtk.gdk.PixbufLoader()
```
and uses the  method:
```
get_pixbuf()
```

to load a GIF icon from a string.

I downgraded back to mingw-w64-x86_64-gdk-pixbuf2 2.36.11-1 and it works again.

The error only happened when running an exe generated by pyinstaller. Running the .py script directly via Python worked fine at all times.

I am using pyinstaller and imageio via pip, and the rest of the dependencies are installed via MSYS2 pacman.

**Comment by lazka** [comment]
Does pyinstaller package all the loaders in /lib/gdk-pixbuf-2.0/2.10.0/loaders/ and their dependencies and does it package the loader cache /lib/gdk-pixbuf-2.0/2.10.0/loaders.cache ?

**Comment by yellowtailfan** [comment]
The only GDK Pixbuf file packaged by pyinstaller is `libgdk_pixbuf-2.0-0.dll`. There are no files with "loader" in the name in the pyinstaller package. It works fine.

This is in the package produced with 2.36.11-1. However, when I had 2.36.11-2 installed I ran a diff on the package contents, and there were some minor differences related to other packages, but nothing to do with GDK Pixbuf.

By the way, I looked in the /mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders folder and in loaders.cache, and none of the loaders mention GIF.

**Comment by lazka** [comment]
There should be "/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-gif.dll" (see https://quodlibet.duckdns.org/msys2/package/mingw-w64-x86_64-gdk-pixbuf2?repo=mingw64)

You need to package those loader*.dll and the cache. Good news is that in MSYS2 the cache is relocatable, so copying should just make it work.

For example svg was always supported through the extra loader, so you were just lucky to not need it until now.

**Comment by yellowtailfan** [comment]
Well, this is odd. I don't have the GIF loader DLL in my MinGW install.
```
λ ls /mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-*.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-ani.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-icns.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-pnm.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-qtif.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-tga.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-xbm.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-xpm.dll
```

This might be an FAQ, but what could cause my install to be missing some of the files?

Meanwhile, I will check if the -2 version of the package has the GIF loader.

**Comment by lazka** [comment]
They are there since 2.36.11-2

**Comment by yellowtailfan** [comment]
OK, the -2 version of the package does include a GIF loader DLL:
```
λ ls /mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-*.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-ani.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-bmp.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-gif.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-icns.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-ico.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-jasper.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-jpeg.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-png.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-pnm.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-qtif.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-tga.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-tiff.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-xbm.dll
/mingw64/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-xpm.dll
```

Perhaps in the previous version, the GIF loader was in the main DLL rather than being in a separate DLL as it appears to be in the -2 version.

**Comment by lazka** [comment]
> Perhaps in the previous version, the GIF loader was in the main DLL rather than being in a separate DLL as it appears to be in the -2 version. That would explain why it worked under pyinstaller without needing to package the GIF loader DLL.

Exactly. I've never used pyinstaller so I don't know what would need to be changed there.. Just copying the files somehow should be enough..

**Comment by yellowtailfan** [comment]
I managed to get the script working from a pyinstaller package with mingw-w64-x86_64-gdk-pixbuf2 2.36.11-2, by adapting the recipe in:
https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Gtk-Application
I also adapted some code that was already in pyinstaller in the files `hook-gi.repository.GdkPixbuf.py` and `pyi_rth_gdkpixbuf.py`. I had to adapt those files to support GTK 2 (as opposed to GTK 3) and MinGW.

It took a while to get this working, as the path construction to the loader DLLs when called from within pyinstaller is very convoluted and sensitive to slash direction and escaping.

Thanks for your help. Unfortunately I'm not in a situation to contribute patches, partly because my solution is still somewhat hard-coded to my script. Hopefully this bug report will be useful if someone else stumbles over this problem in future.
